sharding/shard.go

```diff
@@ -0,0 +1,262 @@
+	}
+	return nil
+}
+
+// HeaderByHash looks up a collation header from the shardDB using the header's hash.
+func (s *Shard) HeaderByHash(hash *common.Hash) (*CollationHeader, error) {
+	encoded, err := s.shardDB.Get(*hash)
+	if err != nil {
+		return nil, fmt.Errorf("get failed: %v", err)
+	}
+	if encoded == nil {
+		return nil, fmt.Errorf("no value set for header hash: %v", hash.String())
+	}
+
+	var header CollationHeader
+
+	stream := rlp.NewStream(bytes.NewReader(*encoded), uint64(len(*encoded)))
+	if err := header.DecodeRLP(stream); err != nil {
+		return nil, fmt.Errorf("could not decode RLP header: %v", err)
+	}
+
+	return &header, nil
+}
+
+// CollationByHash fetches full collation.
+func (s *Shard) CollationByHash(headerHash *common.Hash) (*Collation, error) {
+	header, err := s.HeaderByHash(headerHash)
+	if err != nil {
+		return nil, fmt.Errorf("cannot fetch header by hash: %v", err)
+	}
+
+	body, err := s.BodyByChunkRoot(header.ChunkRoot())
+	if err != nil {
+		return nil, fmt.Errorf("cannot fetch body by chunk root: %v", err)
+	}
+	// TODO: deserializes the body into a txs slice instead of using
+	// nil as the third arg to MakeCollation.
+	col := NewCollation(header, body, nil)
+	return col, nil
+}
+
+// CanonicalHeaderHash gets a collation header hash that has been set as
+// canonical for shardID/period pair.
+func (s *Shard) CanonicalHeaderHash(shardID *big.Int, period *big.Int) (*common.Hash, error) {
+	key := canonicalCollationLookupKey(shardID, period)
+
+	// fetches the RLP encoded collation header corresponding to the key.
+	encoded, err := s.shardDB.Get(key)
+	if err != nil {
+		return nil, err
+	}
+	if encoded == nil {
+		return nil, fmt.Errorf("no canonical collation header set for period=%v, shardID=%v pair: %v", shardID, period, err)
+	}
+
+	// RLP decodes the header, computes its hash.
+	var header CollationHeader
+
+	stream := rlp.NewStream(bytes.NewReader(*encoded), uint64(len(*encoded)))
+	if err := header.DecodeRLP(stream); err != nil {
+		return nil, fmt.Errorf("could not decode RLP header: %v", err)
+	}
+
+	collationHash := header.Hash()
+	return &collationHash, nil
+}
+
+// CanonicalCollation fetches the collation set as canonical in the shardDB.
+func (s *Shard) CanonicalCollation(shardID *big.Int, period *big.Int) (*Collation, error) {
+	h, err := s.CanonicalHeaderHash(shardID, period)
+	if err != nil {
+		return nil, fmt.Errorf("error while getting canonical header hash: %v", err)
+	}
+
+	collation, err := s.CollationByHash(h)
+	if err != nil {
+		return nil, fmt.Errorf("no canonical collation found for hash: %v", err)
+	}
+	return collation, nil
+}
+
+// BodyByChunkRoot fetches a collation body.
+func (s *Shard) BodyByChunkRoot(chunkRoot *common.Hash) ([]byte, error) {
+	body, err := s.shardDB.Get(*chunkRoot)
+	if err != nil {
+		return nil, err
+	}
+	if body == nil {
+		return nil, fmt.Errorf("no corresponding body with chunk root found: %v", chunkRoot.String())
```

Comment:
Change this to `%s` and I _think_ you don't have to call `.String()` since go will do that automatically.